**CLEANING Rates_of_Cases_or_Deaths.csv** (First dataset)

In [1]:
import pandas as pd
import shutil
import re
import numpy as np

In [2]:
raw = pd.read_csv("Data/raw/Rates_of_Cases_or_Deaths.csv")

In [3]:
print(raw)

     outcome     month  MMWR week     Age group Vaccine product  \
0       case  APR 2021     202114         12-17       all_types   
1       case  APR 2021     202114         18-29       all_types   
2       case  APR 2021     202114         30-49       all_types   
3       case  APR 2021     202114         50-64       all_types   
4       case  APR 2021     202114         65-79       all_types   
...      ...       ...        ...           ...             ...   
1586   death  AUG 2022     202235           80+       all_types   
1587   death  AUG 2022     202235  all_ages_adj       all_types   
1588   death  AUG 2022     202235  all_ages_adj         Janssen   
1589   death  AUG 2022     202235  all_ages_adj         Moderna   
1590   death  AUG 2022     202235  all_ages_adj          Pfizer   

      Vaccinated with outcome  Fully vaccinated population  \
0                           8                 3.688700e+04   
1                         674                 2.543093e+06   
2        

HANDLING MISSING AND INVALID DATA

In [4]:
# number of missing values(data)
raw.isnull()

outcome  month  MMWR week  Age group  Vaccine product  \
0       False  False      False      False            False   
1       False  False      False      False            False   
2       False  False      False      False            False   
3       False  False      False      False            False   
4       False  False      False      False            False   
...       ...    ...        ...        ...              ...   
1586    False  False      False      False            False   
1587    False  False      False      False            False   
1588    False  False      False      False            False   
1589    False  False      False      False            False   
1590    False  False      False      False            False   

      Vaccinated with outcome  Fully vaccinated population  \
0                       False                        False   
1                       False                        False   
2                       False                        False   
3                       False                        False   
4                       False                        False   
...                       ...                          ...   
1586                    False                        False   
1587                    False                        False   
1588                    False                        False   
1589                    False                        False   
1590                    False                        False   

      Unvaccinated with outcome  Unvaccinated population  Crude vax IR  \
0                         False                    False         False   
1                         False                    False         False   
2                         False                    False         False   
3                         False                    False         False   
4                         False                    False         False   
...                         ...                      ...           ...   
1586                      False                    False         False   
1587                      False                    False         False   
1588                      False                    False         False   
1589                      False                    False         False   
1590                      False                    False         False   

      Crude unvax IR  Crude IRR  Age adjusted vax IR  Age adjusted unvax IR  \
0              False      False                 True                   True   
1              False      False                 True                   True   
2              False      False                 True                   True   
3              False      False                 True                   True   
4              False      False                 True                   True   
...              ...        ...                  ...                    ...   
1586           False      False                 True                   True   
1587           False      False                False                  False   
1588           False      False                False                  False   
1589           False      False                False                  False   
1590           False      False                False                  False   

      Age adjusted IRR  Continuity correction  
0                 True                  False  
1                 True                  False  
2                 True                  False  
3                 True                  False  
4                 True                  False  
...                ...                    ...  
1586              True                  False  
1587             False                  False  
1588             False                  False  
1589             False                  False  
1590             False                  False  

[1591 rows x 16 columns]

In [5]:
raw.isnull().sum()

outcome                          0
month                            0
MMWR week                        0
Age group                        0
Vaccine product                  0
Vaccinated with outcome          0
Fully vaccinated population      0
Unvaccinated with outcome        0
Unvaccinated population          0
Crude vax IR                     0
Crude unvax IR                   0
Crude IRR                      114
Age adjusted vax IR            987
Age adjusted unvax IR          987
Age adjusted IRR               987
Continuity correction            0
dtype: int64

In [9]:
# % of missing values in each column: number of missing values/total values * 100
raw.isnull().sum()/len(raw)*100

outcome                         0.000000
month                           0.000000
MMWR week                       0.000000
Age group                       0.000000
Vaccine product                 0.000000
Vaccinated with outcome         0.000000
Fully vaccinated population     0.000000
Unvaccinated with outcome       0.000000
Unvaccinated population         0.000000
Crude vax IR                    0.000000
Crude unvax IR                  0.000000
Crude IRR                       7.165305
Age adjusted vax IR            62.036455
Age adjusted unvax IR          62.036455
Age adjusted IRR               62.036455
Continuity correction           0.000000
dtype: float64

In [6]:
## Age adjusted vax IR, Age adjusted unvax IR, and Age adjusted IRR have a high % of missing values. 
## We will delete them. As they are important for the analysis, we will replace them by age group metrics.
clean = raw.drop(["Age adjusted vax IR", "Age adjusted unvax IR", "Age adjusted IRR"], axis=1)
print(clean)

     outcome     month  MMWR week     Age group Vaccine product  \
0       case  APR 2021     202114         12-17       all_types   
1       case  APR 2021     202114         18-29       all_types   
2       case  APR 2021     202114         30-49       all_types   
3       case  APR 2021     202114         50-64       all_types   
4       case  APR 2021     202114         65-79       all_types   
...      ...       ...        ...           ...             ...   
1586   death  AUG 2022     202235           80+       all_types   
1587   death  AUG 2022     202235  all_ages_adj       all_types   
1588   death  AUG 2022     202235  all_ages_adj         Janssen   
1589   death  AUG 2022     202235  all_ages_adj         Moderna   
1590   death  AUG 2022     202235  all_ages_adj          Pfizer   

      Vaccinated with outcome  Fully vaccinated population  \
0                           8                 3.688700e+04   
1                         674                 2.543093e+06   
2        

In [7]:
# Save clean to csv file
clean.to_csv("clean_data_1.csv", index=False)

In [8]:
source = "clean_data_1.csv"
destination = "Data/clean/"
shutil.move(source, destination)

'Data/clean/clean_data_1.csv'

In [18]:
clean = pd.read_csv("Data/clean/clean_data_1.csv")
print(clean)

     outcome     month  MMWR week     Age group Vaccine product  \
0       case  APR 2021     202114         12-17       all_types   
1       case  APR 2021     202114         18-29       all_types   
2       case  APR 2021     202114         30-49       all_types   
3       case  APR 2021     202114         50-64       all_types   
4       case  APR 2021     202114         65-79       all_types   
...      ...       ...        ...           ...             ...   
1586   death  AUG 2022     202235           80+       all_types   
1587   death  AUG 2022     202235  all_ages_adj       all_types   
1588   death  AUG 2022     202235  all_ages_adj         Janssen   
1589   death  AUG 2022     202235  all_ages_adj         Moderna   
1590   death  AUG 2022     202235  all_ages_adj          Pfizer   

      Vaccinated with outcome  Fully vaccinated population  \
0                           8                 3.688700e+04   
1                         674                 2.543093e+06   
2        

In [33]:
## Filling "Crude IRR" as % of missing values is 7.1653:
## Calculate mean of this column and fill the missing values with the mean.
mean_value = clean["Crude IRR"].mean()
clean['Crude IRR'] = clean['Crude IRR'].fillna(mean_value)
print(mean_value)
clean['Crude IRR'].isnull().sum()

7.014692613974949


np.int64(0)

In [35]:
source = "clean_data_1_updated.csv"
destination = "Data/clean/"
shutil.move(source, destination)

'Data/clean/clean_data_1_updated.csv'

In [36]:
clean.isnull().sum()

outcome                        0
month                          0
MMWR week                      0
Age group                      0
Vaccine product                0
Vaccinated with outcome        0
Fully vaccinated population    0
Unvaccinated with outcome      0
Unvaccinated population        0
Crude vax IR                   0
Crude unvax IR                 0
Crude IRR                      0
Continuity correction          0
dtype: int64

In [46]:
# Define a function to check for valid entries in the "outcome" column
def is_valid_outcome(value):
    if pd.isna(value):  # Check for missing values
        return False
    return bool(re.match(r"^(case|death)$", str(value).strip()))  # Normalize value and check match

# Filter invalid entries
invalid_outcome = clean[~clean['outcome'].apply(is_valid_outcome)]

# Count invalid entries
print(f"Number of invalid entries: {len(invalid_outcome)}")

# Remove rows with invalid entries
clean[clean['outcome'].isin(valid_values)].copy()

# Verify the cleaning process
print("Unique values after cleaning:", clean['outcome'].unique())

Number of invalid entries: 0
Unique values after cleaning: ['case' 'death']


In [129]:
# Check for invalid entries in the month column.
# Define a valid pattern for month-year
def is_valid_month(value):
    if pd.isna(value):
        return False
    return bool(re.match(r"^(Jan|Feb|Mar|Apr|May|June|July|Aug|Sept|Oct|Nov|Dec)(2021|2022)$", str(value).strip()))

# Identify invalid entries
invalid_month = clean[~clean['month'].apply(is_valid_month)]

# Display invalid entries
print("Invalid entries in 'month' column:")
print(invalid_month)

# Count invalid entries
print(f"Number of invalid entries: {len(invalid_month)}")

# Remove rows with invalid entries
clean[clean['month'].isin(valid_values)].copy()

# Verify the cleaning process
print("Unique values after cleaning:", clean['month'].unique())

Invalid entries in 'month' column:
     outcome month  MMWR week     Age group Vaccine product  \
0       case   NaT     202114         12-17       all_types   
1       case   NaT     202114         18-29       all_types   
2       case   NaT     202114         30-49       all_types   
3       case   NaT     202114         50-64       all_types   
4       case   NaT     202114         65-79       all_types   
...      ...   ...        ...           ...             ...   
1586   death   NaT     202235           80+       all_types   
1587   death   NaT     202235  all_ages_adj       all_types   
1588   death   NaT     202235  all_ages_adj         Janssen   
1589   death   NaT     202235  all_ages_adj         Moderna   
1590   death   NaT     202235  all_ages_adj          Pfizer   

      Vaccinated with outcome  Fully vaccinated population  \
0                           8                 3.688700e+04   
1                         674                 2.543093e+06   
2                     

In [128]:
clean['month'] = pd.to_datetime(clean['month'], format='%B %Y', errors='coerce')


# Convert the 'month' column with full month names and year to datetime
#clean['month'] = pd.to_datetime(clean['month'], errors='coerce')

# Verify the result
print(clean['month'])

#clean['month'] = pd.to_datetime(clean['month'], format='%m/%Y', errors='coerce')
#print(clean['month'])
#print(clean.columns)

0      NaT
1      NaT
2      NaT
3      NaT
4      NaT
        ..
1586   NaT
1587   NaT
1588   NaT
1589   NaT
1590   NaT
Name: month, Length: 1591, dtype: datetime64[ns]


In [47]:
# checking for invalid age groups
unique_age_groups = clean['Age group'].unique()
print(unique_age_groups)

['12-17' '18-29' '30-49' '50-64' '65-79' '80+' 'all_ages_adj' '5-11']


In [48]:
# valid age group intervals 
expected_age_groups = unique_age_groups = [
    "12-17", "18-29", "30-49", "50-64", "65-79", "80+", "5-11", "all_ages_adj"
]

# Check if all values in the 'Age group' column are valid
unexpected_age_groups = clean[~clean['Age group'].isin(expected_age_groups)]

print("Unexpected Age group values:")
print(unexpected_age_groups['Age group'].unique())

# Number of unexpected age groups
print(f"Number of unexpected age groups: {len(unexpected_age_groups)}")

Unexpected Age group values:
[]
Number of unexpected age groups: 0


In [50]:
clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1591 entries, 0 to 1590
Data columns (total 13 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   outcome                      1591 non-null   object 
 1   month                        1591 non-null   object 
 2   MMWR week                    1591 non-null   int64  
 3   Age group                    1591 non-null   object 
 4   Vaccine product              1591 non-null   object 
 5   Vaccinated with outcome      1591 non-null   int64  
 6   Fully vaccinated population  1591 non-null   float64
 7   Unvaccinated with outcome    1591 non-null   int64  
 8   Unvaccinated population      1591 non-null   float64
 9   Crude vax IR                 1591 non-null   float64
 10  Crude unvax IR               1591 non-null   float64
 11  Crude IRR                    1591 non-null   float64
 12  Continuity correction        1591 non-null   int64  
dtypes: float64(5), int

DUPLICATES

In [51]:
clean.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1586    False
1587    False
1588    False
1589    False
1590    False
Length: 1591, dtype: bool

In [52]:
clean.duplicated().sum()

np.int64(0)

In [54]:
duplicate_count = int(clean.duplicated().sum())
duplicate_count

0

DATETIME

In [55]:
clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1591 entries, 0 to 1590
Data columns (total 13 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   outcome                      1591 non-null   object 
 1   month                        1591 non-null   object 
 2   MMWR week                    1591 non-null   int64  
 3   Age group                    1591 non-null   object 
 4   Vaccine product              1591 non-null   object 
 5   Vaccinated with outcome      1591 non-null   int64  
 6   Fully vaccinated population  1591 non-null   float64
 7   Unvaccinated with outcome    1591 non-null   int64  
 8   Unvaccinated population      1591 non-null   float64
 9   Crude vax IR                 1591 non-null   float64
 10  Crude unvax IR               1591 non-null   float64
 11  Crude IRR                    1591 non-null   float64
 12  Continuity correction        1591 non-null   int64  
dtypes: float64(5), int

In [59]:
clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1591 entries, 0 to 1590
Data columns (total 13 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   outcome                      1591 non-null   object        
 1   month                        0 non-null      datetime64[ns]
 2   MMWR week                    1591 non-null   int64         
 3   Age group                    1591 non-null   object        
 4   Vaccine product              1591 non-null   object        
 5   Vaccinated with outcome      1591 non-null   int64         
 6   Fully vaccinated population  1591 non-null   float64       
 7   Unvaccinated with outcome    1591 non-null   int64         
 8   Unvaccinated population      1591 non-null   float64       
 9   Crude vax IR                 1591 non-null   float64       
 10  Crude unvax IR               1591 non-null   float64       
 11  Crude IRR                    1591 non-null 

In [60]:
# Save the cleaned DataFrame to a CSV file
clean.to_csv('clean_data_1_updated.csv', index=False)

In [62]:
source = "clean_data_1_updated.csv"
destination = "Data/clean/"
shutil.move(source, destination)

'Data/clean/clean_data_1_updated.csv'

In [115]:
clean = pd.read_csv('Data/clean/clean_data_1_updated.csv')

In [116]:
print(clean.columns)

Index(['outcome', 'month', 'MMWR week', 'Age group', 'Vaccine product',
       'Vaccinated with outcome', 'Fully vaccinated population',
       'Unvaccinated with outcome', 'Unvaccinated population', 'Crude vax IR',
       'Crude unvax IR', 'Crude IRR', 'Continuity correction'],
      dtype='object')


**CLEANING County_Level_of_community_Transmission.csv** (second dataset)

In [63]:
raw = pd.read_csv("Data/raw/County_Level_of_community_Transmission.csv")

In [64]:
print(raw)

           state_name        county_name  fips_code        date  \
0               Texas     Jackson County      48239  01/27/2020   
1            Nebraska     Buffalo County      31019  01/27/2020   
2            Illinois       Coles County      17029  01/27/2020   
3                Iowa         Ida County      19093  01/27/2020   
4           Tennessee        Knox County      47093  02/05/2020   
...               ...                ...        ...         ...   
3225217     Louisiana  Assumption Parish      22007  10/18/2022   
3225218  South Dakota        Hyde County      46069  10/18/2022   
3225219         Texas      Loving County      48301  10/18/2022   
3225220      Virginia     Lynchburg City      51680  10/18/2022   
3225221       Wyoming    Converse County      56009  10/18/2022   

        cases_per_100K_7_day_count_change  \
0                                   0.000   
1                                   0.000   
2                                   0.000   
3              

In [65]:
raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3225222 entries, 0 to 3225221
Data columns (total 7 columns):
 #   Column                                              Dtype  
---  ------                                              -----  
 0   state_name                                          object 
 1   county_name                                         object 
 2   fips_code                                           int64  
 3   date                                                object 
 4   cases_per_100K_7_day_count_change                   object 
 5   percent_test_results_reported_positive_last_7_days  float64
 6   community_transmission_level                        object 
dtypes: float64(1), int64(1), object(5)
memory usage: 172.2+ MB


In [66]:
raw.describe()

fips_code  percent_test_results_reported_positive_last_7_days
count  3.225222e+06                                       2.823884e+06 
mean   3.137540e+04                                       1.154418e+01 
std    1.630087e+04                                       1.076036e+01 
min    1.001000e+03                                       0.000000e+00 
25%    1.902900e+04                                       3.790000e+00 
50%    3.002200e+04                                       8.640000e+00 
75%    4.610500e+04                                       1.618000e+01 
max    7.215300e+04                                       1.000000e+02

In [67]:
# raw.describe() does not calculate metrics in column cases_per_100K_7_day_count_change 
# because there are numerical and non-numerical entries in this column.
# so, convert non-numeric to numeric.

# Create a copy of the raw
clean = raw.copy()  
clean['cases_per_100K_7_day_count_change'] = pd.to_numeric(clean['cases_per_100K_7_day_count_change'], errors='coerce')

In [68]:
clean[['cases_per_100K_7_day_count_change', 'percent_test_results_reported_positive_last_7_days']].describe()

cases_per_100K_7_day_count_change  \
count                       2.468979e+06   
mean                        1.898672e+02   
std                         2.016008e+02   
min                        -9.018800e+02   
25%                         3.879500e+01   
50%                         1.277070e+02   
75%                         2.730585e+02   
max                         9.999970e+02   

       percent_test_results_reported_positive_last_7_days  
count                                       2.823884e+06   
mean                                        1.154418e+01   
std                                         1.076036e+01   
min                                         0.000000e+00   
25%                                         3.790000e+00   
50%                                         8.640000e+00   
75%                                         1.618000e+01   
max                                         1.000000e+02

Interpretation

Cases_per_100K_7_day_count_change

count: 2,468,979 valid non-null values. 
mean: The average 7-day case count change per 100,000 people is 189.87. 
std: The standard deviation is 201.60, indicating significant variability in the case count changes. 
min: The lowest change is -901.88 cases, which could indicate a large decrease in cases during a 7-day period. 
Percentiles: 25%: 38.80 (25% of the data is below this value). 50% (median): 127.71 (half the data is below this value). 
             75%: 273.06 (25% of the data is above this value). max: 999.997 cases per 100,000, likely indicating a very high increase.

Percent_test_results_reported_positive_last_7_days

count: 2,823,884 valid non-NaN values. 
mean: The average positivity rate over the last 7 days is 11.54%. 
std: The standard deviation is 10.76%, indicating some variation in positivity rates. 
min: The lowest positivity rate is 0%, indicating areas with no positive cases in the last 7 days. 
Percentiles: 25%: 3.79%. 50% (median): 8.64%. 75%: 16.18%. max: 100%, suggesting some areas are reporting exclusively positive results.

A positivity rate of 100% may indicate areas where testing is limited or only symptomatic individuals are being tested. 
The wide range from 0% to 100% suggests a significant disparity in how COVID-19 impacts different regions.

In [69]:
clean.to_csv("clean_data_2.csv", index=False)

In [86]:
clean = pd.read_csv("Data/clean/clean_data_2.csv")

In [87]:
source = "clean_data_2.csv"
destination = "Data/clean/"
shutil.move(source, destination)

'Data/clean/clean_data_2.csv'

In [71]:
clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3225222 entries, 0 to 3225221
Data columns (total 7 columns):
 #   Column                                              Dtype  
---  ------                                              -----  
 0   state_name                                          object 
 1   county_name                                         object 
 2   fips_code                                           int64  
 3   date                                                object 
 4   cases_per_100K_7_day_count_change                   float64
 5   percent_test_results_reported_positive_last_7_days  float64
 6   community_transmission_level                        object 
dtypes: float64(2), int64(1), object(4)
memory usage: 172.2+ MB


MISSING AND INVALIDE VALUES

In [72]:
clean.isnull().sum()

state_name                                              2662
county_name                                             2662
fips_code                                                  0
date                                                       0
cases_per_100K_7_day_count_change                     756243
percent_test_results_reported_positive_last_7_days    401338
community_transmission_level                           22661
dtype: int64

In [73]:
# % of missing values in each column = number of missing values/total values * 100
clean.isnull().sum()/len(clean) * 100

state_name                                             0.082537
county_name                                            0.082537
fips_code                                              0.000000
date                                                   0.000000
cases_per_100K_7_day_count_change                     23.447781
percent_test_results_reported_positive_last_7_days    12.443733
community_transmission_level                           0.702618
dtype: float64

In [81]:
#Define a function to check for valid entries in the "cases_per_100K_7_day_count_change" column
def is_valid_cases_per_100K_7_day_count_change(value):
    if pd.isna(value):  # Check for missing values
        return False
    try:
        # Try converting the value to float to check if it's a valid number
        float(value)
        return True
    except ValueError:
        # If conversion fails, return False (invalid value)
        return False

# Filter invalid entries
invalid_cases_per_100K_7_day_count_change = clean[~clean['cases_per_100K_7_day_count_change'].apply(is_valid_cases_per_100K_7_day_count_change)]

# Count invalid entries
print(f"Number of invalid entries: {len(invalid_cases_per_100K_7_day_count_change)}")

# Remove rows with invalid entries (Optional step)
clean = clean[clean['cases_per_100K_7_day_count_change'].apply(is_valid_cases_per_100K_7_day_count_change)]

# Verify the cleaning process
print("Unique values after cleaning:", clean['cases_per_100K_7_day_count_change'].unique())


Number of invalid entries: 0
Unique values after cleaning: [  0.          86.683      189.86718457 ... 117.39       154.49
 230.97      ]


In [84]:
#Define a function to check for valid entries in the "percent_test_results_reported_positive_last_7_days" column
def is_valid_percent_test_results_reported_positive_last_7_days(value):
    if pd.isna(value):  # Check for missing values
        return False
    try:
        # Try converting the value to float to check if it's a valid number
        float(value)
        return True
    except ValueError:
        # If conversion fails, return False (invalid value)
        return False

# Filter invalid entries
invalid_percent_test_results_reported_positive_last_7_days = clean[~clean['percent_test_results_reported_positive_last_7_days'].apply(is_valid_percent_test_results_reported_positive_last_7_days)]

# Count invalid entries
print(f"Number of invalid entries: {len(invalid_percent_test_results_reported_positive_last_7_days)}")

# Remove rows with invalid entries (Optional step)
clean = clean[clean['percent_test_results_reported_positive_last_7_days'].apply(is_valid_percent_test_results_reported_positive_last_7_days)]

# Verify the cleaning process
print("Unique values after cleaning:", clean['percent_test_results_reported_positive_last_7_days'].unique())


Number of invalid entries: 0
Unique values after cleaning: [189.86718457   0.          73.33       ...  65.55        71.09
  77.89      ]


In [85]:
clean.to_csv('clean_data_2_updated.csv', index=False)

In [88]:
source = "clean_data_2_updated.csv"
destination = "Data/clean/"
shutil.move(source, destination)

'Data/clean/clean_data_2_updated.csv'

In [92]:

# Define a function to check for valid entries in the "outcome" column
def is_valid_community_transmission_level(value):
    if pd.isna(value):  # Check for missing values
        return False
    return bool(re.match(r"^(high|low|moderate|substantial)$", str(value).strip()))  # Normalize value and check match

# Filter invalid entries
invalid_community_transmission_level = clean[~clean['community_transmission_level'].apply(is_valid_community_transmission_level)]

# Count invalid entries
print(f"Number of invalid entries: {len(invalid_community_transmission_level)}")

# Remove rows with invalid entries
clean[clean['community_transmission_level'].isin(valid_values)].copy()

# Verify the cleaning process
print("Unique values after cleaning:", clean['community_transmission_level'].unique())


Number of invalid entries: 22661
Unique values after cleaning: [nan 'low' 'high' 'moderate' 'substantial']


In [93]:
mode_value = clean['community_transmission_level'].mode().iloc[0]
clean['community_transmission_level'] = clean['community_transmission_level'].fillna(mode_value)

In [94]:
print("Unique values after cleaning:", clean['community_transmission_level'].unique())

Unique values after cleaning: ['high' 'low' 'moderate' 'substantial']


In [95]:
# Define a function to check for valid entries in the "outcome" column
def is_valid_state_name(value):
    if pd.isna(value):  # Check for missing values
        return False
    return bool(re.match(r"^(value)$", str(value).strip()))  # Normalize value and check match

# Filter invalid entries
invalid_state_name = clean[~clean['state_name'].apply(is_valid_state_name)]

# Count invalid entries
print(f"Number of invalid entries: {len(invalid_state_name)}")

# Remove rows with invalid entries
clean[clean['state_name'].isin(valid_values)].copy()

# Verify the cleaning process
print("Unique values after cleaning:", clean['state_name'].unique())

Number of invalid entries: 3225222
Unique values after cleaning: ['Texas' 'Nebraska' 'Illinois' 'Iowa' 'Tennessee' 'Wisconsin' 'Missouri'
 'New York' 'Alabama' 'Ohio' 'Utah' 'Georgia' 'Minnesota' 'Alaska'
 'New Jersey' 'Colorado' 'South Dakota' 'West Virginia' 'California'
 'Michigan' 'Kansas' 'Nevada' 'Montana' 'Kentucky' 'Oregon' 'Indiana'
 'Mississippi' 'Virginia' 'Idaho' 'South Carolina' 'North Carolina'
 'Oklahoma' 'Florida' 'Hawaii' 'North Dakota' 'Arkansas' 'Maine'
 'Pennsylvania' 'Vermont' 'Wyoming' 'Puerto Rico' 'Arizona' 'New Mexico'
 'Washington' 'Louisiana' 'Connecticut' 'Maryland' nan
 'District of Columbia' 'Delaware' 'Massachusetts' 'Rhode Island'
 'New Hampshire']


In [98]:
mode_value = clean['state_name'].mode().iloc[0]
clean['state_name'] = clean['state_name'].fillna(mode_value)
print("Unique values after cleaning:", clean['state_name'].unique())

Unique values after cleaning: ['Texas' 'Nebraska' 'Illinois' 'Iowa' 'Tennessee' 'Wisconsin' 'Missouri'
 'New York' 'Alabama' 'Ohio' 'Utah' 'Georgia' 'Minnesota' 'Alaska'
 'New Jersey' 'Colorado' 'South Dakota' 'West Virginia' 'California'
 'Michigan' 'Kansas' 'Nevada' 'Montana' 'Kentucky' 'Oregon' 'Indiana'
 'Mississippi' 'Virginia' 'Idaho' 'South Carolina' 'North Carolina'
 'Oklahoma' 'Florida' 'Hawaii' 'North Dakota' 'Arkansas' 'Maine'
 'Pennsylvania' 'Vermont' 'Wyoming' 'Puerto Rico' 'Arizona' 'New Mexico'
 'Washington' 'Louisiana' 'Connecticut' 'Maryland' 'District of Columbia'
 'Delaware' 'Massachusetts' 'Rhode Island' 'New Hampshire']


In [99]:
clean.to_csv('clean_data_2_updated.csv', index=False)

In [100]:
source = "clean_data_2_updated.csv"
destination = "Data/clean/"
shutil.move(source, destination)

'Data/clean/clean_data_2_updated.csv'

CONVERTING DATE

In [108]:
# Format the 'date' column to 'm/y'
clean['date'] = clean['date'].dt.strftime('%m/%Y')

# Verify the result
print(clean['date'])

0          01/2020
1          01/2020
2          01/2020
3          01/2020
4          02/2020
            ...   
3225217    10/2022
3225218    10/2022
3225219    10/2022
3225220    10/2022
3225221    10/2022
Name: date, Length: 3225222, dtype: object


CHECKING DUPLICATES

In [105]:
clean.duplicated().sum()
duplicat_count = int(clean.duplicated().sum())
duplicat_count

0